### PR TITLE
Johnson Algorithm implementation in C (Issue #2206)

### LINKED ISSUE #2206: Johnson Algorithm
- [x] C 
- [ ] C++ - @Aakansha99
- [x] C# - @ritish099
- [ ] CoffeeScript
- [x] Dart - @Swarnimashukla
- [ ] Go - @BhardwajShweta25
- [ ] Java - @Neelaksh-Singh
- [ ] JS - @thakran14
- [x] Kotlin - @ritish099
- [x] Php - @ritish099
- [ ] Python - @BishalKrChanda
- [ ] Readme - @AdityaSingh17
- [ ] Ruby
- [ ] Rust
- [ ] TypeScript

### PR DESCRIPTION
**Pull request by ka1shi**
Fixes #2206

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Addition of folder for Johnson Algorithm
- Addition of code for Johnson Algorithm in C

#### Languages Used:
- C

#### Files Added:
- Johnson_Algorithm.c

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            // Check if old distance is greater than new distance
+            if(dist[u]!=INT_MAX && dist[v]>dist[u]+w)
+            {
+                // Update old distance with new distance
+                dist[v] = dist[u]+w;
+            }
+        }
+    }
+    
+    // Information Message
```
Comment (jainaman224) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Remove this commented section.

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+Distance from 6 to 4 is :6
+Distance from 6 to 5 is :6
+Distance from 6 to 6 is :0
+
+        
+ */
+
+
+
+
```
Comment (jainaman224) [inline Johnson_Algorithm/Johnson_Algorithm.c]: There should be a single new line at the end of file.

File: Johnson_Algorithm/Johnson_Algorithm.c (hunk 1 above)
Comment (ka1shi) [reply Johnson_Algorithm/Johnson_Algorithm.c]: Done @jainaman224

File: Johnson_Algorithm/Johnson_Algorithm.c (hunk 2 above)
Comment (ka1shi) [reply Johnson_Algorithm/Johnson_Algorithm.c]: Done @jainaman224

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+    struct Edge* edge;
+};
+
+struct Graph* createGraph(int V, int E)
+{
+    // Create a graph with V vertices and E Edges
+    
+    // Allocate memory to V vertices of graph
+    struct Graph* graph = malloc(V*sizeof(int));
+    graph->V = V;
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around ->

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    graph->E = E;
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around ->

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+    
+    // Allocate memory to E Edges of graph
+    graph->edge = malloc(E*sizeof(int));
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around the operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    // Information Message
+    //printf("The graph is created with %d vertices and %d edges.\n", V, E);
+    return graph;
+}
+
+struct Graph* addEdge(struct Graph* graph, int src, int des, int wt, int i)
+{
+    // Add Edges with source, destination and weight
+    
+    graph->edge[i].source = src;
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    return graph;
+}
+
+void printGraph(struct Graph* graph)
+{
+    // Print graph as distance of connected vertex with each other
+    
+    int E = graph->E;
+    
+    for(int i=0; i<E; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        printf("The distance from %d to %d is: %d\n",graph->edge[i].source, graph->edge[i].destination, graph->edge[i].weight);
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around the operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    printf("\n");
+}
+
+int* BellmanFord(struct Graph* graph, int extra, int* dist)
+{
+    // Shortest distance from extra vertex to every vertex in graph
+    
+    int V=graph->V;
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+    int E=graph->E;
+    
+    // Initialize distance from extra vertex to all vertices as Infinite(INT_MAX)
+    for(int i=0; i<V+1; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+        }
+    }
+    
+    printf("Final Result\n");
+    printf("(Value 2147483648 indicates vertex v is not reachable from vertex u)\n\n");
+    for(int i=0; i<V; i++)
+    {
+        for(int j=0; j<V; j++)
+        {
+            if(distance[i][j]==INT_MAX)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+        {
+            distance[i][j] = dist_final[j];
+        }
+    }
+    
+    printf("Final Result\n");
+    printf("(Value 2147483648 indicates vertex v is not reachable from vertex u)\n\n");
+    for(int i=0; i<V; i++)
+    {
+        for(int j=0; j<V; j++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+    
+    
+    // New graph with all positive weight distance
+    //printf("Updated Graph:\n\n");
+    //printGraph(graph);
+    
+    //Matrix to store distance of every vertex with every other vertex
+    int distance[V][V];
+    //Running Dijsktra for every vertex as source
+    for(int i=0; i<V; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around the operator. Do this in the whole file.

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+        // Update the final matrix distance
+        for(int j=0; j<V; j++)
+        {
+            distance[i][j] = dist_final[j];
+        }
+    }
+    
+    printf("Final Result\n");
+    printf("(Value 2147483648 indicates vertex v is not reachable from vertex u)\n\n");
+    for(int i=0; i<V; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+{
+    // Create a graph with V vertices and E Edges
+    
+    // Allocate memory to V vertices of graph
+    struct Graph* graph = malloc(V*sizeof(int));
+    graph -> V = V;
+    graph -> E = E;
+    
+    // Allocate memory to E Edges of graph
+    graph -> edge = malloc(E*sizeof(int));
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around *

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    return graph;
+}
+
+void printGraph(struct Graph* graph)
+{
+    // Print graph as distance of connected vertex with each other
+    
+    int E = graph -> E;
+    
+    for(int i=0; i<E; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+
+int* BellmanFord(struct Graph* graph, int extra, int* dist)
+{
+    // Shortest distance from extra vertex to every vertex in graph
+    
+    int V=graph -> V;
+    int E=graph -> E;
+    
+    // Initialize distance from extra vertex to all vertices as Infinite(INT_MAX)
+    for(int i=0; i<V+1; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 18]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        dist[i] = INT_MAX;
+    }
+    
+    // Distance from extra vertex to itself is zero
+    dist[extra] = 0;
+    
+    // Shortest path from any vertex can have atmost V-1 edges
+    for(int i=1; i<=V; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 19]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        for(int j=0; j<E+V; j++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 20]
```diff
@@ -0,0 +1,371 @@
...
+        {
+            int u = graph -> edge[j].source;
+            int v = graph -> edge[j].destination;
+            int w = graph -> edge[j].weight;
+            
+            // Check if old distance is greater than new distance
+            if(dist[u]!=INT_MAX && dist[v]>dist[u]+w)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 21]
```diff
@@ -0,0 +1,371 @@
...
+    }
+    
+    return dist;
+}
+
+struct Graph* updateEdge(struct Graph* graph, int dist[], int E)
+{
+    // Update Edge after getting distance from extra vertex to each original vertex
+    // To make distance between all vertices poitive
+    for(int i=0; i<E; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 22]
```diff
@@ -0,0 +1,371 @@
...
+}
+
+int minDistance(int dist[], bool visited[], int V)
+{
+    // Find closest vertex from source vertex that has not been visited
+    
+    // Initialize as min as INFINITY(INT_MAX)
+    int min = INT_MAX, min_index;
+    
+    for(int v=0; v<V; v++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+        }
+    }
+    return min_index;
+}
+
+int askGraph(struct Graph* graph, int u, int v, int E)
+{
+    // Return distance from vertex u to vertex v
+    
+    for(int i=0; i<E; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+
+int* Dijsktra(struct Graph* graph, int V, int E, int src, int* dist)
+{
+    // Calculate shortest path from source vertex to all other vertices
+    
+    // checks if vertex i is visited and is finalized
+    bool visited[V]; 
+  
+    // Initialize all distances as INT_MAX and visited[] as false 
+    for(int i=0; i<V; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 25]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        dist[i] = INT_MAX;
+        visited[i] = false;
+    }
+    
+    // Distance of source vertex from itself is zero
+    dist[src] = 0;
+    
+    // Shortest path from any vertex can have atmost V-1 edges
+    for(int i=0; i<V-1; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 26]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        //Pick the minimum distance vertex from all vertices
+        int u = minDistance(dist, visited, V);
+        
+        //Picked vertex is visited
+        visited[u] = true;
+        
+        //Update distance of adjacent vertices from picked vertex
+        for(int v=0; v<V; v++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 27]
```diff
@@ -0,0 +1,371 @@
...
+    scanf("%d %d", &V, &E);
+    
+    // Information Message
+    //printf("%d\n",V);
+    //printf("%d\n",E);
+    
+    // Declaring source, destination and weight for edges
+    int src, des, wt;
+    
+	// (V+1) One Extra vertex is considered for calculating distance with each vertex present in original graph
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: inconsistency in indentation

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 28]
```diff
@@ -0,0 +1,371 @@
...
+    // Declaring source, destination and weight for edges
+    int src, des, wt;
+    
+	// (V+1) One Extra vertex is considered for calculating distance with each vertex present in original graph 
+	// Additional V edges (E+V) from extra vertex to all original vertices is considered
+	
+    struct Graph* graph = createGraph(V+1, E+V); 
+    
+    // Add each edge to graph with source, destination and weight
+    for(int i=0; i<E+V; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 29]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        if(i<E) // Original E Edges
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c]: Add space around operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 30]
```diff
@@ -0,0 +1,371 @@
...
+    int V, E;
+    
+    // Array of Edge structure
+    struct Edge* edge;
+};
+
+struct Graph* createGraph(int V, int E)
+{
+    // Create a graph with V vertices and E Edges
+
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:27]: Remove this line

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 31]
```diff
@@ -0,0 +1,371 @@
...
+    
+    // Information Message
+    // printf("The graph is created with %d vertices and %d edges.\n", V, E);
+    return graph;
+}
+
+struct Graph* addEdge(struct Graph* graph, int src, int des, int wt, int i)
+{
+    // Add Edges with source, destination and weight
+
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:44]: Remove this line

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 32]
```diff
@@ -0,0 +1,371 @@
...
+{
+    // Add Edges with source, destination and weight
+    
+    graph -> edge[i].source = src;
+    graph -> edge[i].destination = des;
+    graph -> edge[i].weight = wt;
+    
+    // Information Message
+    // printf("Edge from %d to %d with distance %d is added.\n",src, des, wt);
+
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:51]: Remove this line

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 33]
```diff
@@ -0,0 +1,371 @@
...
+    // Information Message
+    // printf("Edge from %d to %d with distance %d is added.\n",src, des, wt);
+    
+    return graph;
+}
+
+void printGraph(struct Graph* graph)
+{
+    // Print graph as distance of connected vertex with each other
+
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:58]: Remove this line

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 34]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        printf("The distance from %d to %d is: %d\n",graph -> edge[i].source, graph -> edge[i].destination, graph -> edge[i].weight);
+    }
+    printf("\n");
+}
+
+int* BellmanFord(struct Graph* graph, int extra, int* dist)
+{
+    // Shortest distance from extra vertex to every vertex in graph
+
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:71]: Remove this line

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 35]
```diff
@@ -0,0 +1,371 @@
...
+void printGraph(struct Graph* graph)
+{
+    // Print graph as distance of connected vertex with each other
+    
+    int E = graph -> E;
+    
+    for(int i = 0; i < E; i++)
+    {
+        printf("The distance from %d to %d is: %d\n",graph -> edge[i].source, graph -> edge[i].destination, graph -> edge[i].weight);
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:63]: Leave space after ,

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 36]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        for(int j = 0; j < V; j++)
+        {
+            if(distance[i][j] == INT_MAX)
+            {
+                distance[i][j] = INT_MAX;
+                // To indicate that j vertex is not reachable from i vertex
+                // No Edge from i to j
+            }
+            printf("Distance from %d to %d is :%d\n",i, j, distance[i][j]);
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:290]: Add a space after ,

**Comment by ka1shi** [comment]
Done @somya-kapoor Kindly review again.

**Comment by somya-kapoor** [comment]
There are 2 files in the PR, please remove the unwanted file. Also, squash the commits

**Comment by ka1shi** [comment]
Hi @somya-kapoor, I followed these steps --> **https://github.com/wprig/wprig/wiki/How-to-squash-commits** to squash the commits. But it somehow result in this "ka1shi force-pushed the ka1shi:Johnson_Algo_C branch from d07b952 to 1d4b2ac" Please do tell me how it should be done or what else could be done. Thank You!

**Comment by somya-kapoor** [comment]
Try http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request

**Comment by ka1shi** [comment]
@somya-kapoor Thanks a lot. It's done. Kindly check.

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 37]
```diff
@@ -0,0 +1,359 @@
+// Implementation of Johnson's algorithm in C
+
+#include<stdio.h>
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:3]: Add Space before <

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 38]
```diff
@@ -0,0 +1,359 @@
...
+#include<limits.h>
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:4]: Add Space before <

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 39]
```diff
@@ -0,0 +1,359 @@
...
+#include<stdlib.h>
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:5]: Add Space before <

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 40]
```diff
@@ -0,0 +1,359 @@
...
+    // printf("%d\n",V);
+    // printf("%d\n",E);
+    
+    // Declaring source, destination and weight for edges
+    int src, des, wt;
+    
+	// One Extra vertex (V+1) is considered for calculating distance with each vertex present in original graph 
+	// Additional V edges (E+V) from extra vertex to all original vertices is considered
+	
+    struct Graph* graph = createGraph(V+1, E+V);
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:208]: Add space around the operator

File: Johnson_Algorithm/Johnson_Algorithm.c [hunk 41]
```diff
@@ -0,0 +1,359 @@
...
+    printf("\n");
+}
+
+int* BellmanFord(struct Graph* graph, int extra, int* dist)
+{
+    int V = graph -> V;
+    int E = graph -> E;
+    
+    // Initialize distance from extra vertex to all vertices as Infinite(INT_MAX)
+    for(int i = 0; i < V+1; i++)
```
Comment (somya-kapoor) [inline Johnson_Algorithm/Johnson_Algorithm.c:65]: Add space around the operator

**Comment by somya-kapoor** [comment]
After making the changes please squash the commits, then it will ready to be merged.

**Comment by ka1shi** [comment]
@somya-kapoor Done! Thanks!

**Comment by somya-kapoor** [comment]
@ka1shi Did you ran your code because I am getting an error while running it.

**Comment by ka1shi** [comment]
@somya-kapoor Let me check. I am working on it.

**Comment by ka1shi** [comment]
@somya-kapoor I run the code on Codechef IDE and it is running completely fine. Can you share the name of the compiler or IDE where you ran the code? Thanks!

**Comment by somya-kapoor** [comment]
It's working fine, my mistake.

**Pull request merged**